File: src/Test/PolymorphismFFI.js
const intMonoidish = {
    mempty_: 1,
    mappend_: function(x) {
      return function(y) {
        return x + y;
      };
    }
};

function polyLoop(dict, n, acc) {
  while (n !== 0) {
    acc = dict.mappend_(acc)(dict.mempty_);
    n--;
  }
  return acc;
}

export const runPolymorphismFFI = function(limit) {
  return polyLoop(intMonoidish, Math.floor(limit), 0);
};

File: src/Test/PolymorphismFFI.go
package Test_PolymorphismFFI

type Monoidish[A any] struct {
	Mempty  A
	Mappend func(A) func(A) A
}

func polyLoop[A any](dict Monoidish[A], n int, acc A) A {
	for n != 0 {
		acc = dict.Mappend(acc)(dict.Mempty)
		n--
	}
	return acc
}

var intMonoidish = Monoidish[int]{
	Mempty: 1,
	Mappend: func(x int) func(int) int {
		return func(y int) int { return x + y }
	},
}

func RunPolymorphismFFI(limit int) int {
	return polyLoop(intMonoidish, limit, 0)
}
